benchmarks/NN/functional_c#/Program.cs

```diff
@@ -0,0 +1,152 @@
+            var correctPredictions = results.Where(res => res.prediction == res.actual).Count();
+            var accuracy = correctPredictions / (double)results.Length * 100.0;
+            
+            System.Console.WriteLine(accuracy);
+        }
+
+        private static ImmutableArray<Layer> backwardsPropagateError(ImmutableArray<Layer> network, ImmutableArray<double> expected)
+        {
+            var firstLayer = network.First();
+            var remainingLayers = network.TakeLast(network.Length - 1);
+            var errors = firstLayer.Zip(expected, (neuron, exp) => exp - neuron.output).ToImmutableArray();
+            
+            Func<double, double> derivative = (neuronOutput) => neuronOutput * (1.0 - neuronOutput);
+            Func<double, Neuron, Neuron> getError = (error, neuron) => new Neuron(neuron.weights, neuron.output, neuron.bias, error * derivative(neuron.output));
+            Func<ImmutableArray<double>, Layer, Layer> delta = (errors, layer) => errors.Zip(layer, (error, neuron) => getError(error, neuron)).ToImmutableArray();
+            Func<Layer, int, double> neuronError = (layer, i) => layer.Select(neuron => neuron.weights[i] * neuron.delta).Sum();
+            Func<ImmutableArray<Layer>, Layer, ImmutableArray<Layer>> backProp = (res, layer) => {
+                var errors = layer.Select((neuron, i) => neuronError(res[0], i)).ToImmutableArray();
+                var newLayer = delta(errors, layer);
+                return res.Add(newLayer);
+            };
+
+            return remainingLayers.Aggregate(ImmutableArray<Layer>.Empty.Add(delta(errors, firstLayer)), (acc, r) => backProp(acc, r)).Reverse().ToImmutableArray();
+        }
+
+        private static (ImmutableArray<Layer>, ImmutableArray<double>) forwardPropagate(ImmutableArray<double> row, ImmutableArray<Layer> network)
+        {
+            Func<ImmutableArray<double>, ImmutableArray<double>, double, double> activate = (weights, inputs, bias) => bias + Enumerable.Zip(weights, inputs, (w, i) => w * i).Sum(); //TODO: Aggregate with bias as seed?
+            Func<double, double> transfer = (activation) => 1.0 / (1.0 + Math.Exp(-activation));
+            Func<ImmutableArray<double>, Neuron, Neuron> propagateNeuron = (inputs, neuron) => new Neuron(neuron.weights, transfer(activate(neuron.weights, inputs, neuron.bias)), neuron.bias, neuron.delta);
+            Func<Layer, ImmutableArray<double>> getOutput = (layer) => layer.Select(neuron => neuron.output).ToImmutableArray();
+
+            var propagatedLayersTemp = network.Aggregate((row, layers: ImmutableArray<Layer>.Empty), (acc, layer) => {
+                var newLayer = layer.Select(neuron => propagateNeuron(acc.Item1, neuron)).ToImmutableArray();
+                return (getOutput(newLayer), acc.Item2.Add(newLayer));
+            });
+            var propagatedLayers = (propagatedLayersTemp.row, propagatedLayersTemp.layers.Reverse().ToImmutableArray());

```

Comment:
Når du alligevel laver det at vende netværket om som et reverse kald, burde kaldet så ikke være i backwardsPropagate, da det er den funktion der har brug for det?